Overhaul process reaping and terminal control

mqudsi opened the pull request:
I _believe_ this finally safely does away with the keepalive process. I think the problem is that we were waiting on all child processes in `process_mark_finished_children` which caused premature reaping of terminated processes before their process group could be joined by new processes in the same job.

Additionally, there was a deadlock in `try_select` that didn't show up until that change was made.

This should fix #4178 (which manifests in two *entirely* different ways depending on whether connected over SSH and FreeBSD vs Linux) and quite likely some other issues as well.

~~This branch should likely be squash-merged.~~

---

Partially addresses #206.

Closes #3805. Closes #3952. Closes #4178. Closes #4238. Closes #4929. Closes #5210.

mqudsi commented:
Also fixes #3952 :)

mqudsi commented:
Also fixes #3805

mqudsi commented:
Also fixes #5210

mqudsi commented:
@ridiculousfish @krader1961 any thoughts?

I've installed this branch as my main driver on my FreeBSD, WSL, and now macOS as it appears to include some highly platform-specific behavior. Would appreciate if others could try it out for a few days, you never know where edge cases can pop up.

faho commented:
>Instead of reaping all child processes when we receive a SIGCHLD, try
reaping only processes belonging to process groups from
fully-constructed jobs.

This basic idea seems entirely logical to me, and I struggle to think how that would make it any worse. Are there cases where that now doesn't reap a process at all? Is it a problem to leave unreaped jobs around for a bit?

I don't think so, but this stuff is full of weird edges and spooky corners.

------------------

I've done some cursory testing, and I can confirm that #5210 seems to be solved. I can't confirm #3952, and that should still  be an issue.

The case there is:

```fish
sudo echo lol | grep lol
```

with `grep` being a function. The big problem is that that is two separate jobs, because fish forgets the current job as soon as it executes the function. Which is something separate from the keepalive stuff, but not made worse by it either.

faho commented:
I can confirm #3805 as fixed, pgroup-wise.

mqudsi commented:
There is no problem in keeping the zombie processes unreaped until the job is constructed, but the problem is whether or not the reaping function is executed at the right places, because when a process is reaped it changes the internal state of the job, and `select_try` would hang when it is waiting for IO that never comes becomes and it deadlocks waiting for the job to change state (which never happens).

----

That's weird, I tested #3952 it yesterday (it even came up in the autosuggest from history) but you're right, it is indeed not fixed.

```fish
mqudsi@ZBook /m/c/U/m/r/fish-shell> function grep
                                        command grep --color=auto
                                    end
mqudsi@ZBook /m/c/U/m/r/fish-shell> sudo echo lol | grep lol
Usage: grep [OPTION]... PATTERN [FILE]...
Try 'grep --help' for more information.
mqudsi@ZBook /m/c/U/m/r/fish-shell> sudo echo lol | grep lol
Usage: grep [OPTION]... PATTERN [FILE]...
Try 'grep --help' for more information.
```

faho commented:
>That's weird, I tested #3952 it yesterday (it even came up in the autosuggest from history) but you're right, it is indeed not fixed.

There is a timing component to it, so it's possible you just got unlucky and got a false-positive.

mqudsi commented:
I also changed the debug levels of some tracing output so that `-d3` can be used for sanely debugging job status and progression.

faho commented:
>I also changed the debug levels of some tracing output so that -d3 can be used for sanely debugging job status and progression.

Ooh yes. Last I looked, the most spammy part was input - "checking X against mapping Y" or something - but I think I already increased the level on that.

I still believe something like #4818 or a file-based thing (`fish --debug=input,reader` to get messages from input.cpp and reader.cpp) would be better, but we're a bit too chatty in low levels in general.

mqudsi commented:
😄 Yeah, it was the key binding mapping that would print page after page of output.

faho commented on src/input.cpp:
```diff
@@ -384,7 +384,7 @@ static bool input_mapping_is_match(const input_mapping_t &m) {
 
     assert(str.size() > 0 && "zero-length input string passed to input_mapping_is_match!");
-    debug(4, L"trying to match mapping %ls", escape_string(m.seq.c_str(), ESCAPE_ALL).c_str());
+    debug(8, L"trying to match mapping %ls", escape_string(m.seq.c_str(), ESCAPE_ALL).c_str());
```
Honestly: Just remove this one. There's not been a bug here for quite a while, and we wouldn't use this debug to catch it.

faho commented on src/reader.cpp:
```diff
@@ -1743,8 +1743,17 @@ static void reader_interactive_init() {
         for (unsigned long loop_count = 0;; loop_count++) {
             pid_t owner = tcgetpgrp(STDIN_FILENO);
-            shell_pgid = getpgrp();
+            // 0 is a valid return code from `tcsetpgrp()` under FreeBSD, and testing
```
in which case... why doesn't mine return 0?

It says it "will return the foreground process group ID of that terminal if there is one, and some value larger than 1 that is not presently a process group ID otherwise". Which just strikes me as quite stupid, and just asking for TOCTTOUs. Why not just 0?

mqudsi replied:
Yup. Thanks.

mqudsi replied:
0 is actually a valid pid for the top-level process (`init` under Linux, `kernel` under FreeBSD).

Under Linux in my test cases I was getting -1, but when SSH'd to FreeBSD I was getting zero. The FreeBSD man page has this to say:

```
     The tcgetpgrp() function returns the value of the process group ID of the foreground process group
     associated with the terminal device.  If there is no foreground process group, tcgetpgrp() returns an
     invalid process ID.
```

A process gets assigned a parent id of zero when its own parent dies (giving parentship of the process to init, whose job is to then call `waitpid` to kill the zombie process). I'm not sure if that's what was happening in my test case (or how that could be what was happening in my case, and I'm suspecting that FreeBSD is returning zero to imply invalid.

I should clarify that comment.

faho replied:
>0 is actually a valid pid for the top-level process (init under Linux, kernel under FreeBSD).

Init is always PID 1 in linux and [FreeBSD](https://www.freebsd.org/doc/handbook/basics-processes.html). PID 0 on linux is the swapper and scheduler, but those don't have children or terminals, so we shouldn't ever have to care about them. I'm not sure if they actually "do processy stuff" on BSD.

>A process gets assigned a parent id of zero when its own parent dies

On linux it gets 1, as above. And I'm pretty sure that applies to BSD too, since I'm pretty sure that also reparents to init.

mqudsi replied:
Sorry, I had the pid for init wrong. FreeBSD does reparent to 1 as well. Which makes this even more bizarre, and lends credence to the theory that 0 does indeed represent invalid under FreeBSD.. yet indicates that a subsequent attempt at `tcsetpgrp` will succeed.

mqudsi commented on src/proc.cpp:
```diff
@@ -264,6 +264,12 @@ int job_signal(job_t *j, int signal) {
 
+static void mark_job_complete(const job_t *j) {
+    for (auto &p : j->processes) {
+        p->completed = 1;
```
Yeah, c boolean that I should probably just change over to a real `bool`.

mqudsi commented on src/proc.cpp:
```diff
@@ -789,7 +823,7 @@ bool terminal_give_to_job(const job_t *j, bool cont) {
     // been done.
-    if (tcgetpgrp(STDIN_FILENO) == j->pgid) {
+    if (j->pgid == -2 || tcgetpgrp(STDIN_FILENO) == j->pgid) {
         debug(4, L"Process group %d already has control of terminal\n", j->pgid);
```
I'm not sure that's true. If a job was all functions or builtins that didn't fork, pgid should be `-2`, but since it's actually operating under the scope of the current shell instance, it technically has control of the shell, no?

faho replied:
Shouldn't the pgid then be the pgid of the shell?

AFAIK, "-2" here is the sentinel value for "hasn't been set up yet".

mqudsi replied:
I'm pretty sure the code that sets the pgid isn't called since iirc it's basically set on the first forked command in the chain. If there was no pgid-related code, -2 is indeed a reflection on the fact that at no point was a pgid assigned.

I'd have to double-check, though.

mqudsi replied:
builtins that do not fork keep pgrp -2.

mqudsi commented on src/postfork.cpp:
```diff
@@ -137,7 +137,7 @@ bool maybe_assign_terminal(const job_t *j) {
 
     if (j->get_flag(JOB_TERMINAL) && j->get_flag(JOB_FOREGROUND)) {  //!OCLINT(early exit)
-        if (tcgetpgrp(STDIN_FILENO) == j->pgid) {
+        if (j->pgid == -2 || tcgetpgrp(STDIN_FILENO) == j->pgid) {
```
You're right, I didn't catch that assert (was working on results of a find operation).

mqudsi replied:
I imagine while 1 is a valid pgid that there's no (useful?) way for a command executed under fish to end up with that pgid?

faho replied:
I think the only way is "init=/bin/fish" on the kernel commandline. Which might be useful.

mqudsi commented on src/parse_execution.cpp:
```diff
@@ -1223,6 +1223,7 @@ parse_execution_result_t parse_execution_context_t::run_1_job(tnode_t<g::job> jo
         // Check to see if this contained any external commands.
         bool job_contained_external_command = false;
         for (const auto &proc : job->processes) {
+            /* debug(1, L"Parsed job with head %ls", proc->argv0()); */
```
I'll keep it but bump it up to level 4. It's the clearest way of seeing what job was parsed first in a pipeline.

faho replied:
Okay!

mqudsi commented on src/reader.cpp:
```diff
@@ -3318,7 +3329,9 @@ const wchar_t *reader_readline(int nchars) {
     if (!reader_exit_forced()) {
-        if (tcsetattr(0, TCSANOW, &old_modes) == -1) {
+        // The order of the two conditions below is important. Try to restore the mode
+        // in all cases, but only complain if not interactive.
```
That should be "if interactive"

ridiculousfish commented:
I'd be totally thrilled to eliminate the keepalive process. I hope this works!!

I think I understand the theory of this PR. Before we would create a keep-alive process, and kill it once the job is constructed. With this PR, we no longer construct a keep-alive. Rather we allow the pgroup to become a zombie, and just delay reaping it by never calling waitpid(), only waiting on certain processes. That's pretty clever. I understand old WSL did not support this (zombies were not a WSL thing), but this was recently improved to make zombie processes real, so this might really work.

The tricky part is that waitpid() is not like select(): you can wait for THIS PARTICULAR child, or wait for ANY child, but not wait for a *set* of children. If you disallow waiting for ANY child, you might wind up waiting on a long-lived background process and thereby hang. So how do we ensure we aren't waiting for a background process?

I think the reason this works is as follows: parser_t puts new jobs at the beginning of the job list, and there can only be one constructed foreground job. We only call wait() after adding a foreground job, so we never end up waiting on a background job. So that's working by accident  the job list happens to put new jobs at the beginning. Yet I think this has hit onto something really great.

I'd like to see the following:

1. `process_mark_finished_children` currently takes one parameter, `wants_await`. Let's change it to `block_on_foreground`. If that's set, this happens:
  a. It asserts that there's at most one job that's foreground AND constructed.
  b. If there is such a job, wait on that job with WNOHANG = 0 (i.e. allow it to block).
  c. For every background job, wait with WNOHANG set.

2. This PR currently has 9 commits, some of which seem to be fixes for previous commits. Can you please clean this up into reviewable chunks? For example I see commented-out code like `// needs_keepalive = true`, we shouldn't have that sort of thing (we shouldn't have `needs_keepalive` at all if this works!) `git rebase -i`  is a good tool for this sort of cleanup.

3. There's multiple uses of `-2` to mean a not-yet constructed job, I think it's time to introduce a constant for this!

4. `proc_wait_any()` invoked from `builtin_wait` is going to be a problem. In particular if you call builtin `wait` from within a function and the function is in a pipeline, you may end up reaping the keepalive process. (We'll want to add a test case for this.)

mqudsi commented:
I ended up spending pretty much the entire day on this. I didn't expect it to be that hard after the work that was already done, but I ran into some corner cases with `waitpid` (documented in the comments.. extensively) that made me pull out half my hair. I've based and rebased more times than I can count, but I think I have the branch in a state where it can (after review) be merged without squashing. Each commit should individually pass the tests.

The most important (non-aesthetic) changes are in 525f329 and  f987183. The changes in 6add7d are small, but do have significant ramifications to error handling in the `job_continue` code.

@ridiculousfish apologies in advance for 4eb01c6, but please see the commit log. Others needing this treatment include the unnamed return result containing the `STATUS_*` values (so that one can distinguish between functions returning an actual int, e.g. 'int count_active_blah_blah_blah` vs those return a status (e.g. `exec_subshell`) and those returning a boolean masquerading as an int (some of which have been replaced with the new `Tristate` which is very similar to `maybe_t<bool>`). 

I introduced RAII wrappers in places to clean up error handling, though there was one place where I was forced to use a lambda as the cleanup depends on which parts of the function were successfully executed.

Debug level 3 generally means "be verbose about job execution" while level 4 means "include details about waits and io handling". 

Most dead code was trimmed from `exec.cpp` and (particularly) `proc.cpp`. The overall flow in `proc.cpp` should be significantly easier to follow, and code has been grouped together in a more-C++ less-C fashion there. In general, self-documenting code styles have been used where possible. Lots more refactoring is needed for the codebase as a whole to take advantage of some of these idioms, but it's not relevant to this branch and shouldn't be committed here.

I tried to keep both code and functional testing in mind when breaking up the changes into commits. I hope I have made things fairly easy to follow and provided sufficient commentary for anyone trying to puzzle something out, now or in the future.

Feedback is most welcome.

Now excuse me while I go forget all the useless things I learned today.

mqudsi commented:
@ridiculousfish I replied there,  it shouldn't be a problem. Do you have any thoughts on the actual logic changes in this PR?

mqudsi commented:
@ridiculousfish https://github.com/mqudsi/fish-shell/blob/keep_zombies/src/enum_set.h

ridiculousfish commented on src/proc.h:
```diff
@@ -192,6 +192,8 @@ class job_t {
     /// Sets the command.
     void set_command(const wcstring &cmd) { command_str = cmd; }
 
+    const wcstring preview() const { return processes.empty() ? L"" : processes[0]->argv0() + wcstring(L" ..."); }
```
First `const` is unnecessary. Should be `wcstring preview() const { ...} 

Can you add a comment here describing what this is for?

ridiculousfish commented on src/common.h:
```diff
@@ -31,6 +31,16 @@
 #include "maybe.h"
 #include "signal.h"  // IWYU pragma: keep
 
+// some versions of ncurses globally define TRUE/FALSE
```
We shouldn't need to be doing this; please remove these defines and TRUE/FALSE definitions.

ridiculousfish commented on src/common.h:
```diff
@@ -978,6 +988,32 @@ enum {
     STATUS_INVALID_ARGS = 121,
 };
 
+/// Replaces usages of int-as-bool to incorporate minus-one-on-failure behavior with a strictly-typed
```
This seems like an unnecessary abstraction to me. I don't think we should introduce this Tristate type. It's not motivated by this commit.

ridiculousfish commented on src/proc.h:
```diff
@@ -139,23 +140,23 @@ typedef std::unique_ptr<process_t> process_ptr_t;
 
 /// Constants for the flag variable in the job struct.
-enum job_flag_t {
+enum class JobFlag {
```
fish style is snake_case not CamelCase; this should be `job_flag_t`.

ridiculousfish commented on src/enum_set.h:
```diff
@@ -0,0 +1,46 @@
+#include <bitset>
+
+template<typename T>
+class EnumSet {
```
Ok I like this idea! I think we may even want to add bitwise operator overloads later.

ridiculousfish commented on src/enum_set.h:
```diff
@@ -0,0 +1,46 @@
+class EnumSet {
+private:
+	std::bitset<8 * sizeof(typename std::underlying_type<T>::type)> bitmask {0};
+	int index_of(T t) const;
```
`index_of` should be `static`. BTW you'll probably find it a lot more pleasant to define all these methods inline:

    static int index_of(T t) {
        static_cast<typename std::underlying_type<T>::type>(t);
    }

No more duplicated templates, etc.

ridiculousfish commented on src/exec.cpp:
```diff
@@ -985,22 +985,17 @@ static bool exec_process_in_job(parser_t &parser, process_t *p, job_t *j,
 }
 
 void exec_job(parser_t &parser, job_t *j) {
+    assert(j != nullptr);
```
for new assertions we should add an "error message:"

     assert(j != nullptr && "Job should not be null");

ridiculousfish commented on src/signal.h:
```diff
@@ -40,4 +40,17 @@ bool signal_is_blocked();
+	}
+
+	~signal_block_t() {
+		signal_unblock();
```
Indentation looks off here; please use `./build_tools/style.fish` on every change.

ridiculousfish commented on src/enum_set.h:
```diff
@@ -0,0 +1,46 @@
+
+template<typename T>
+class EnumSet {
+private:
```
I think a helpful typedef is warranted here:

    using base_type_t = std::underlying_type<T>::type;

to avoid repeating `std::underlying_type<T>::type`.

ridiculousfish commented on src/proc.cpp:
```diff
@@ -681,8 +684,9 @@ void proc_update_jiffies() {
-static int select_try(job_t *j) {
+/// \return Tristate::yes() if buffers were ready, Tristate::no() if no buffers were ready,
+/// and Tristate::err() on error.
+static Tristate select_try(job_t *j) {
```
Rather than this Tristate abstraction which IMO won't prove useful enough to justify, let's use an `enum class` here:

    enum class select_return_t {
          some_readable,   // some fd was readable
          none_readable,  // no fd was readable
          no_fds,  // no fds were present in the IO chain
    };

(probably you can think of better value names).

I think this will be more explicit.

ridiculousfish commented:
Overall really nice work! My comments are mostly nits.

Did you want to get this into fish 3? My feeling is we should put it in the release after, this is a big change.

mqudsi commented:
Thanks for reviewing, you basically poked at everything I was feeling "meh" about.

I would prefer not to include this in fish 3.0.

faho commented:
>I would prefer not to include this in fish 3.0.

That would mean either putting off #5210 or finding some other, less invasive, way to fix it.

mqudsi commented:
I guess it depends on how much testing this would see if it were included in fish 3.0, which depends on a) how many people are running off of `master` and b) when 3.0 will be frozen.

zanchey commented:
I'd be happy to take it for the beta as that is likely to get more widespread testing.

mqudsi commented on src/proc.cpp:
```diff
@@ -363,96 +363,78 @@ typedef unsigned int process_generation_count_t;
+            // No killed/dead children in this particular process group
+        } else {
+            // pid < 0 indicates an error. One likely failure is ECHILD (no children), which we
+            // break on, and is not considered an error. The other likely failure is EINTR,
```
Bad comment, actual logic was correct.

mqudsi commented:
@faho #3952 kept eating at me, 30dbdb0 should take care of it? I'm unsure how much of this is a hack and how much of it would ultimately be required anyway.

mqudsi commented:
Now also closes #4238.

faho commented:
I'm not quite sure what it is, but this feels really sluggish to use?

Typing is slower, the git prompt takes twice as long.

I've not been able to find a clean reproducer so far unfortunately.

mqudsi commented:
@faho was it like that the list time you tried?

Thanks to all the rebasing, tracking it down to a particular change is going to be a PITA 😐

faho commented:
>was it like that the list time you tried?

I don't _think_ so, but I've not specifically been on the lookout.

So far, it seems like there's multiple slowdowns here. Up to eb166b265 (the second commit), this behaves like master (~55ms). Then 4fa24b3fe slows down to ~85ms, and the full monty (550ecc5ec) goes down to ~190ms.

That's just running `fish_right_prompt` and `echo $cmd_duration` interactively.

Profiling it (via `fish --profile`) shows that e.g. `set -l changedFiles (command git diff --name-status 2>/dev/null | string match -r \\w)` slowed down from 4ms to 14ms. The command substitution is necessary.

So far the smallest reproducer I have found is:

```fish
set -l foo (command true | string match .)
```

This goes from 1ms to 11ms. Using `command true | true` is not enough.

mqudsi commented:
It occurs to me that there's a chance `cmd_duration`'s calculation changed in the process. I would rely on an external tool (e.g. `bash -c 'time fish -c "foo"'` rather than fish's own metrics for this.

faho commented:
>The first commit you mentioned 4fa24b3 was almost entirely aesthetic with the exception of using a bitset to store the job flags.

That's from eb166b2 to 4fa24b3, so including the preceding commits, including e.g. 05bcd0195. I didn't test every commit, only these (random-ish) points.

If you want, I can check more, but I felt like a nice reproducer was more important.

>It occurs to me that there's a chance cmd_duration's calculation changed in the process.

It's also noticeable, especially since my prompt is also affected, so I get quite a noticeable effect from prompt to prompt.

mqudsi commented:
> If you want, I can check more, but I felt like a nice reproducer was more important.

Ah, thanks. That makes more sense. fwiw, I test via `git bisect` so that helps pin it down without having to try all the commits.

>It's also noticeable, especially since my prompt is also affected, so I get quite a noticeable effect from prompt to prompt.

I was just thinking aloud about other possible reasons why those commits had the problem, but your explanation makes that a moot point.

mqudsi commented:
The entire design of `proc.cpp` makes it an impossible choice between working correctly or working fast. I pushed an update that tried to minimize the `waitpid` calls, but the architecture there is just wrong and the impact is still felt.

You don't timeout the `select(2)` calls every 10 milliseconds to check if processes have exited, then try to make that finish as quickly as possible so you can get back to processing IO. You just don't.

Waiting on processes to exit and waiting on IO to become available are two top-level operations that need to be run concurrently. Calling `waitpid(-1, ...)` enabled the use of `SIGCHLD` counting to determine if we call `waitpid` in the first place, but it was 100% wrong because posix wasn't architectured that way. Now we need to make several transitions between user code and system code because we can't wait on all processes in one go (there's no way to specify multiple pids or pgrps to wait on any of them).

The correct approach here would be either a separate thread with synchronization allowing the main thread to push the pgrps of constructed jobs to it, and this thread would be responsible for waiting on them to exit and setting their status correctly or a thread spawned per job (which is much cleaner) to wait on the job's children to exit and set their status accordingly. "Setting process status" is already (mostly) thread safe since at some point (long ago) it used to be done in a signal handler (probably on SIGCHLD directly). Or a separate thread to block until IO is ready and then handle the results.

Doing this doesn't just get rid of the slowdown from switching to `waitpid` by pgrp: it's even faster and less taxing than the old code since there's no constant timing out from the `select` call just to see if we need to handle the exiting of any processes.

I'm not sure what to do.

ridiculousfish commented:
I agree with all this, though in fish's defense the "wait 10 msec" path is only in semi-weird cases involving redirections between external commands and fish functions; it's not the case for most executions.

The right thing to do here is whatever helps prepare us for to concurrent execution, which I think we can achieve in 3.1. Here's a sketch of how I expect it to work:

1. When a fish function (or certain builtins) is executed in a pipeline, a new exec thread is created for the function. This exec thread is conceptually analogous to a forked child. The function performs synchronous I/O (`read()`, `write()`) on this thread.

2. The parent thread's job is to wait for its "processes" some of which may be fish exec threads, and some of which are external. 

In this model, the parent thread no longer has to perform I/O. `select_try()` will just go away. `job_continue()` will do the following:

1. While I have a foreground (non-background) external command, `waitpid()` it.
2. Otherwise, wait for child exec threads (lots of possible mechanisms here).
3. Opportunistically reap background jobs.

Note that multiple threads will be running `job_continue()` at once, each waiting for their own children.

This PR moves us closer to this model since we no longer wait on all processes. So let's not tackle the `select_try()` issue yet since we're going to solve that via a big architectural shift.

mqudsi commented:
So to be clear, your suggestion is that this PR is fine (any other issues notwithstanding) for now?

faho commented:
>The right thing to do here is whatever helps prepare us for to concurrent execution, which I think we can achieve in 3.1.

The question I have is what we should do for 3.**0**?

We shouldn't release with #5210, but the perf here is bad enough that I would not want to release with this solution (which is a really nice piece of work, and seems to solve most problems with job control! Great job @mqudsi!).

Would it be possible to merge just part of this? Specifically up to eb166b265fb181ce979de01b0e552fb0665603e4, since 514eea9358f3d8b15cbd4d5fb25fd47edcc3619e is the really juicy part, and that has basically 0 performance impact.

Any further than that and it gets quite bad and noticeable. Note that we're not talking about a couple of percent here, this is 30% on the git prompt, and probably most completions.

mqudsi commented:
Yeah, I guess I wasn't clear. I know what should be done, just not sure what should be done now. I guess now that I have a blessing to introduce threads into fish core I can try to hack something in temporarily, but I'm not sure if there's a place where that's possible without ending up with the rewrite.

mqudsi commented:
@faho would you be so kind as to test it again? I found a (fragile and convoluted) approach that lets us use the SIGCHLD count to elide waitpid calls under certain cases.

faho commented:
>My expect test script is running ~20% faster under keep_zombies than under master.

I can confirm that! Very nice!

faho commented:
However, this is missing a `#include <functional>` in common.h for me (gcc 8.2.1). You might want to add that!

faho commented on src/proc.cpp:
```diff
@@ -395,35 +428,60 @@ static bool process_mark_finished_children(bool block_on_fg) {
+        // cases, calling waitpid(2) is faster than potentially calling select_try() on a process
+        // that has exited, which will force us to wait the full timeout before coming back here and
+        // calling waitpid() again.
+        while (true) {
```
Isn't this a bit clearer as `while (auto pid = waitpid`?

faho replied:
(I'm not a massive fan of `while (true)` because that means you need to manually account for all possibilities to not have an infinite loop)

mqudsi replied:
Changed. 

rust's preference for the unconditional `loop` over `while ...` has gotten to me, and I found it more logical to explicitly "handle" all possible results of the `waitpid` call instead of split the handling across the while condition and inside the body.

faho replied:
AFAIK rust has compiler support for checking that you handled all possible results. This is C++, which... doesn't.

mqudsi replied:
Very fair point!

faho commented on src/proc.cpp:
```diff
@@ -395,35 +428,60 @@ static bool process_mark_finished_children(bool block_on_fg) {
         }
     }
 
+    // Yes, the below can be collapsed to a single line, but it's worth being explicit about it with
```
I _would_ actually shorten this to:

```c++
// If we skipped any jobs, we can't definitely say whether all children were reaped.
// If we didn't, we can just return next time if we got no SIGCHLD (i.e. fresh children to reap).
dirty_state = jobs_skipped;
```

mqudsi replied:
Very well then! :)

faho commented on src/proc.cpp:
```diff
@@ -362,27 +362,60 @@ static bool process_mark_finished_children(bool block_on_fg) {
+    /// Tracks whether or not we received SIGCHLD without checking all jobs (due to jobs under
+    /// construction), forcing a full waitpid loop.
+    static bool dirty_state = true;
+    static size_t last_jobs_hash = -1;
```
Is this used anywhere?

mqudsi replied:
Nope. As mentioned, that entire commit was part of a different optimization strategy and has now been dropped via rebase.

faho commented on src/proc.h:
```diff
@@ -316,6 +319,9 @@ class job_iterator_t {
 
+    /// Returns a hash of all jobs currently running that can be used to determine if jobs have
+    /// changed. Does not reflect changes to the states of these jobs.
+    size_t hash() const;
```
Where is this used?

mqudsi replied:
No, sorry that entire commit should be dropped. I used to use a snapshot of the jobs to determine if we can safely elide waitpid(2) calls, but that logic has been changed now.

faho commented on src/postfork.cpp:
```diff
@@ -63,39 +63,56 @@ static void debug_safe_int(int level, const char *format, int val) {
-            safe_perror("setpgid");
-            retval = false;
+
+        while (int result = setpgid(p->pid, j->pgid) != 0) {
```
This groups like

```
(int result = setpgid()) != 0
```

and not like

```
int result = (setpgid() != 0)
```

right?

(I can never recall operator precedence so I just add parens everywhere)

mqudsi replied:
Good catch.

mqudsi replied:
This was actually even more broken than that, I was testing the 0/1 `result` instead of `errno` 🤦‍♂️

faho commented on src/postfork.cpp:
```diff
@@ -63,39 +63,56 @@ static void debug_safe_int(int level, const char *format, int val) {
+                // with tcsetpgrp(2) in other places and it disappears on retry.
+                debug_safe(2, "setpgid(2) returned EPERM. Retrying");
+            }
+            else if (result == EINTR) {
```
Don't we usually let the "else" ride on the closing bracket?

```c++
} else if
```

?

mqudsi replied:
I know we do with final `} else {` clauses, but I wasn't sure about `} else if (...) {` but a quick check with `ag '\}\n\s*else if' -c` reveals you are correct.

faho commented on src/exec.cpp:
```diff
@@ -996,6 +998,32 @@ void exec_job(parser_t &parser, job_t *j) {
+    // write to downstream of a pipe.
+    // By keeping track of (select) "jobs in flight" we can try to marshall newly-created external
+    // processes into existing pgrps.
+    static std::stack<job_t *> active_jobs{};
```
This entire thing strikes me as a bit of a hack, to be honest.

(The "real" solution is probably to pass the job along in exec_job, but that then means passing it all through the parser and that's just annoying)

I _think_ it works (and it seems to in actual tests), though.

Is there any way that a job in active_jobs becomes stale?

mqudsi replied:
It absolutely is a hack. But the entire parser approach is to set up a new context for the evaluation of functions and I broke too many things when I tried to change that.

`active_jobs` only contains `job_t` instances that haven't yet left `exec_job`, to keep track of the parent in event of recursive `exec_job` calls due to evaluation of functions and blocks. Jobs are removed from `active_jobs` before `exec_job` ends, so they're guaranteed to always be still "in execution" and very much valid.

I *can* change `exec_job` to take a `shared_ptr<job_t>` since that's what we actually originally have then decompose it to to a raw pointer for some reason, but there's really no need, it's guaranteed safe.

mqudsi replied:
I added a HACK warning and explanation of the preferred route plus complications you'll encounter going that way.

faho commented on src/exec.cpp:
```diff
@@ -1031,43 +1031,6 @@ void exec_job(parser_t &parser, job_t *j) {
         }
     }
 
-    // When running under WSL, create a keepalive process unconditionally if our first
```
There is absolutely no way somebody is running "old WSL"?

mqudsi replied:
Of course there is. But are we going to burden our codebase forever with junk working around bugs while WSL reaches maturity? It's a tradeoff between when fish 3.0 gets and released and adoption of the latest WSL releases as time goes on, and it's easy enough to tell users running into pgrp-related errors (`tcsetpgrp`/`setpgid`, primarily) under WSL to update their Windows.

Keep in mind that WSL isn't on by default under Windows 10 and it's only the most tech savvy Windows users that will know what WSL is and use it.

faho replied:
I actually did not know that! I thought it was an actual thing now, with releases and such!

faho commented on src/proc.cpp:
```diff
@@ -515,6 +515,8 @@ static int process_clean_after_marking(bool allow_interactive) {
 
             // TODO: The generic process-exit event is useless and unused.
             // Remove this in future.
+            // Update: This event is used for cleaning up the psub temporary files and folders.
```
Huh? Is this the same event as the "--on-job-exit caller" event?

mqudsi replied:
I know, I was confused too. `PROCESS_EXIT` is fired from two places in the codebase and I can't see where it's used. This PR is already getting out of hand so I don't want to figure out why it breaks the job-exit event/the psub cleanup, but it does.

faho replied:
Oh, yeah, sure, by all means keep this as-is!

faho commented on src/proc.cpp:
```diff
@@ -702,20 +703,20 @@ static int select_try(job_t *j) {
-        tv.tv_sec = 0;
-        tv.tv_usec = 10000;
+        timeout.tv_sec = 0;
+        timeout.tv_usec = 10000;
```
Btw: Why is this 10ms? Even millliseconds is really quite long in computer time, so wouldn't e.g. 1ms suffice?

faho replied:
This used to be 10 _seconds_ from the initial revision up to 6b8e7b16f6d4e11e168e3ce2effe2d8f0a53b184.

https://sourceforge.net/p/fish/mailman/message/23293756/ is the mail that included the patch, and the justification for 10ms seems to be "shorter than 5s, still long enough that it's not busy-waiting". But 10ms still seems to be _really_ long.

mqudsi replied:
Good job digging that up. 10ms is long, but 1ms timeouts have a noticeable CPU cost. I would rather just leave this at 10ms for now until we do the concurrency changes we were discussing, at which point the timeout can become `-1` and fish can properly react to both IO and child process exits without polling.

faho commented on src/proc.cpp:
```diff
@@ -794,12 +791,16 @@ bool terminal_give_to_job(const job_t *j, bool cont) {
         // block indefinitely).
-        while (tcsetpgrp(STDIN_FILENO, j->pgid) != 0) {
-            debug(3, "tcsetpgrp failed");
+        while (true) {
```
I actually prefer the old loop condition.

mqudsi replied:
I agree. I think during refactoring and rebasing a billion times over the final code works by simply plugging `tcsetpgrp` back into the condition but at some point in the middle that wasn't possible. Reverted.

faho commented on src/proc.cpp:
```diff
@@ -754,19 +753,17 @@ static void read_try(job_t *j) {
-    signal_block();
+    // RAII wrappers must have a name so that their scope is tied to the function as it is legal for
+    // the compiler to construct and then immediately deconstruct unnamed objects otherwise.
+    signal_block_t signal_block;
```
This stuff is nice!

But let me just say that I think whoever named a bunch of C++ idioms has no naming skills whatsoever. RAII is, as a term, crap.

mqudsi replied:
😄 what would you do if I threw in a SFINAE or two in there?

faho commented on src/proc.h:
```diff
@@ -11,7 +11,7 @@
 #include <unistd.h>
 
-#include <list>
+#include <deque>
```
Have you tried vector? Any benchmarks?

From all the benchmarks I've seen, list is pretty much useless for almost all workloads (cache misses were the main issue IIRC), but vector is often nice.

See e.g. https://baptiste-wicht.com/posts/2012/12/cpp-benchmark-vector-list-deque.html and https://www.youtube.com/watch?v=YQs6IC-vgmo.

Also if both performed the same we should prefer vector because we're already using it.

mqudsi replied:
`deque` is like a `vector` but with more efficient insert at the head too, not just the end. All our inserts were to the top of the list, so we should have actually changed this to a vector *plus* changed our view of the jobs list from top-to-bottom to bottom-to-top (just using `rbegin()` and `rend()` in iteration would have sufficed), but we can do that some other time.

faho replied:
Yeah, fair enough!

faho commented on src/proc.cpp:
```diff
@@ -363,96 +363,80 @@ typedef unsigned int process_generation_count_t;
 /// the SIGCHLD signal handler, and therefore does not need atomics or locks.
 static volatile process_generation_count_t s_sigchld_generation_cnt = 0;
```
If we're not using it anymore, why is it still there?

mqudsi replied:
We're using it now.

faho replied:
Ah, sorry, I was going commit-for-commit, and the message for a5a23c57d82faa60a00dd6f7cd618a80c92c7c7d ("Overhaul behavior of `process_mark_finished_children()`") says:

>We're not using the SIGCHLD count in s_sigchld_generation_cnt for
anything any more, as it's not actually a reliable metric

mqudsi replied:
Understood. I actually debated deleting it in that same commit but left it around to see if we would need it later (which proved to be the case). I don't think I want to merge a5a23c5 with the waitpid optimizations commit, I would rather they be two separate, testable/bisectable chunks.

faho replied:
Okay.

mqudsi commented:
Thanks for testing and reviewing, @faho. I appreciate the confirmation about the performance concerns being put to rest (thank God!) and the other feedback.

> However, this is missing a #include <functional> in common.h for me (gcc 8.2.1). You might want to add that!

It's odd that [travis didn't catch that](https://travis-ci.org/fish-shell/fish-shell/builds/440904110?utm_source=github_status&utm_medium=notification), but I've fixed it in the latest rebase, along with the other issues you pointed out.

faho commented:
>It's odd that travis didn't catch that

Travis is basically always on the edge of Ubuntu EOL. Don't expect new compiler versions there!

faho commented:
@mqudsi: There's a weird error message in the tests that I don't believe was there before - https://travis-ci.org/fish-shell/fish-shell/jobs/443932043#L1131:

>killpg: Invalid argument

If I'm reading this correctly, this happens in tests/interactive.fish before the "Testing interactive functionality" message is printed.

faho commented:
Yup, `source`ing test_util.fish STOPs in a weird state (PID = 0?), after

```fish
    exec ../test/root/bin/fish $script $args_for_test_script
```

It seems running `exec` in a script will do that.

mqudsi commented:
Interesting. It only happens with a literal `source`.

`foo.fish` contains the following: `exec build/fish -c "echo hello"`

`fish foo.fish` runs  fine. But `fish -c 'source foo.fish` prints "killpg: invalid argument".

(There's no actual difference between those two commands, is there?)

faho commented:
Okay, I got a nice reproducible crash here:

```fish
echo a >somefile
grep a somefile | grep c | grep d
```
results in:

```
<W> fish: Could not send job 2 ('command grep --color=auto $argv') with pgid 14699 to foreground
tcsetpgrp: Kein passender Prozess gefunden
<W> fish: Could not send job 2 ('command grep --color=auto $argv') with pgid 14699 to foreground
tcsetpgrp: Kein passender Prozess gefunden
<E> fish: /tmp/makepkg/fish-shell-git/src/fish-shell/src/proc.cpp:411: failed assertion: job_fg == nullptr && "More than one active, fully-constructed foreground job!"
<E> fish: Backtrace:
<E> fish: 0   job_reap(bool) + 0
<E> fish: 1   job_t::continue_job(bool) + 1032
<E> fish: 2   exec_job(parser_t&, job_t*) + 7028
<E> fish: 3   parse_execution_context_t::run_1_job(tnode_t<grammar::job>, block_t const*) + 3156
<E> fish: 4   parse_execution_context_t::run_job_conjunction(tnode_t<grammar::job_conjunction>, block_t const*) + 196
<E> fish: 5   parse_execution_result_t parse_execution_context_t::run_job_list<grammar::job_list>(tnode_t<grammar::job_list>, block_t const*) + 193
<E> fish: 6   parse_execution_context_t::eval_node(tnode_t<grammar::job_list>, block_t const*, io_chain_t const&) + 1380
<E> fish: 7   int parser_t::eval_node<grammar::job_list>(std::shared_ptr<parsed_source_t const>, tnode_t<grammar::job_list>, io_chain_t const&, block_type_t) + 392
<E> fish: 8   parser_t::eval(std::shared_ptr<parsed_source_t const>, io_chain_t const&, block_type_t) + 146
<E> fish: 9   parser_t::eval(std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, io_chain_t const&, block_type_t) + 224
<E> fish: 10  /usr/bin/fish(+0xaae43) [0x560f269c0e43]
<E> fish: 11  /usr/bin/fish(+0xaf557) [0x560f269c5557]
<E> fish: 12  /usr/bin/fish(+0xafee8) [0x560f269c5ee8]
<E> fish: 13  expand_string(std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, std::vector<completion_t, std::allocator<completion_t> >*, int, std::vector<parse_error_t, std::allocator<parse_error_t> >*) + 647
<E> fish: 14  parse_execution_context_t::expand_arguments_from_nodes(std::vector<tnode_t<grammar::argument>, std::allocator<tnode_t<grammar::argument> > > const&, std::vector<std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, std::allocator<std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > > >*, parse_execution_context_t::globspec_t) + 450
<E> fish: 15  parse_execution_context_t::populate_plain_process(job_t*, process_t*, tnode_t<grammar::plain_statement>) + 1011
<E> fish: 16  parse_execution_context_t::populate_job_process(job_t*, process_t*, tnode_t<grammar::statement>) + 231
<E> fish: 17  parse_execution_context_t::populate_job_from_job_node(job_t*, tnode_t<grammar::job>, block_t const*) + 322
<E> fish: 18  parse_execution_context_t::run_1_job(tnode_t<grammar::job>, block_t const*) + 1053
<E> fish: 19  parse_execution_context_t::run_job_conjunction(tnode_t<grammar::job_conjunction>, block_t const*) + 196
<E> fish: 20  parse_execution_context_t::run_if_statement(tnode_t<grammar::if_statement>, block_t const*) + 315
<E> fish: 21  parse_execution_context_t::run_1_job(tnode_t<grammar::job>, block_t const*) + 2739
<E> fish: 22  parse_execution_context_t::run_job_conjunction(tnode_t<grammar::job_conjunction>, block_t const*) + 196
<E> fish: 23  parse_execution_result_t parse_execution_context_t::run_job_list<grammar::job_list>(tnode_t<grammar::job_list>, block_t const*) + 193
<E> fish: 24  parse_execution_context_t::eval_node(tnode_t<grammar::job_list>, block_t const*, io_chain_t const&) + 1380
<E> fish: 25  int parser_t::eval_node<grammar::job_list>(std::shared_ptr<parsed_source_t const>, tnode_t<grammar::job_list>, io_chain_t const&, block_type_t) + 392
<E> fish: 26  void internal_exec_helper<grammar::job_list>(parser_t&, std::shared_ptr<parsed_source_t const>, tnode_t<grammar::job_list>, io_chain_t const&) + 194
<E> fish: 27  exec_job(parser_t&, job_t*) + 8134
<E> fish: 28  parse_execution_context_t::run_1_job(tnode_t<grammar::job>, block_t const*) + 1870
<E> fish: 29  parse_execution_context_t::run_job_conjunction(tnode_t<grammar::job_conjunction>, block_t const*) + 196
<E> fish: 30  parse_execution_result_t parse_execution_context_t::run_job_list<grammar::job_list>(tnode_t<grammar::job_list>, block_t const*) + 193
<E> fish: 31  parse_execution_context_t::eval_node(tnode_t<grammar::job_list>, block_t const*, io_chain_t const&) + 1380
<E> fish: 32  int parser_t::eval_node<grammar::job_list>(std::shared_ptr<parsed_source_t const>, tnode_t<grammar::job_list>, io_chain_t const&, block_type_t) + 392
<E> fish: 33  parser_t::eval(std::shared_ptr<parsed_source_t const>, io_chain_t const&, block_type_t) + 146
<E> fish: 34  parser_t::eval(std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, io_chain_t const&, block_type_t) + 224
<E> fish: 35  /usr/bin/fish(+0xaae43) [0x560f269c0e43]
<E> fish: 36  /usr/bin/fish(+0xaf557) [0x560f269c5557]
<E> fish: 37  /usr/bin/fish(+0xafee8) [0x560f269c5ee8]
<E> fish: 38  expand_string(std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, std::vector<completion_t, std::allocator<completion_t> >*, int, std::vector<parse_error_t, std::allocator<parse_error_t> >*) + 647
<E> fish: 39  parse_execution_context_t::expand_arguments_from_nodes(std::vector<tnode_t<grammar::argument>, std::allocator<tnode_t<grammar::argument> > > const&, std::vector<std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, std::allocator<std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > > >*, parse_execution_context_t::globspec_t) + 450
<E> fish: 40  parse_execution_context_t::populate_plain_process(job_t*, process_t*, tnode_t<grammar::plain_statement>) + 1011
<E> fish: 41  parse_execution_context_t::populate_job_process(job_t*, process_t*, tnode_t<grammar::statement>) + 231
<E> fish: 42  parse_execution_context_t::populate_job_from_job_node(job_t*, tnode_t<grammar::job>, block_t const*) + 322
<E> fish: 43  parse_execution_context_t::run_1_job(tnode_t<grammar::job>, block_t const*) + 1053
<E> fish: 44  parse_execution_context_t::run_job_conjunction(tnode_t<grammar::job_conjunction>, block_t const*) + 196
<E> fish: 45  parse_execution_result_t parse_execution_context_t::run_job_list<grammar::job_list>(tnode_t<grammar::job_list>, block_t const*) + 193
<E> fish: 46  parse_execution_context_t::eval_node(tnode_t<grammar::job_list>, block_t const*, io_chain_t const&) + 1380
<E> fish: 47  int parser_t::eval_node<grammar::job_list>(std::shared_ptr<parsed_source_t const>, tnode_t<grammar::job_list>, io_chain_t const&, block_type_t) + 392
<E> fish: 48  void internal_exec_helper<grammar::job_list>(parser_t&, std::shared_ptr<parsed_source_t const>, tnode_t<grammar::job_list>, io_chain_t const&) + 194
<E> fish: 49  exec_job(parser_t&, job_t*) + 8134
<E> fish: 50  parse_execution_context_t::run_1_job(tnode_t<grammar::job>, block_t const*) + 1870
<E> fish: 51  parse_execution_context_t::run_job_conjunction(tnode_t<grammar::job_conjunction>, block_t const*) + 196
<E> fish: 52  parse_execution_result_t parse_execution_context_t::run_job_list<grammar::job_list>(tnode_t<grammar::job_list>, block_t const*) + 193
<E> fish: 53  parse_execution_context_t::eval_node(tnode_t<grammar::job_list>, block_t const*, io_chain_t const&) + 1380
<E> fish: 54  int parser_t::eval_node<grammar::job_list>(std::shared_ptr<parsed_source_t const>, tnode_t<grammar::job_list>, io_chain_t const&, block_type_t) + 392
<E> fish: 55  parser_t::eval(std::shared_ptr<parsed_source_t const>, io_chain_t const&, block_type_t) + 146
<E> fish: 56  parser_t::eval(std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, io_chain_t const&, block_type_t) + 224
<E> fish: 57  /usr/bin/fish(+0xaae43) [0x560f269c0e43]
<E> fish: 58  /usr/bin/fish(+0xf7f73) [0x560f26a0df73]
<E> fish: 59  reader_readline(int) + 171
<E> fish: 60  reader_read(int, io_chain_t const&) + 525
<E> fish: 61  main + 4064
<E> fish: 62  __libc_start_main + 243
<E> fish: 63  _start + 46
```

mqudsi commented:
I'm presuming grep is a function?

Are you on the latest version of the code? I fixed that a few days ago, it was caused by sharing pgrps across function jobs.

faho commented:
>I'm presuming grep is a function?

Yup, the standard `--color=auto` thing.

>Are you on the latest version of the code?

It turns out I wasn't. I checked github, but that says you added commits "22 days ago" and "3 days ago" respectively. It does not like this rebasing business.

Could you therefore post a quick comment that you added something when you do?

>I fixed that a few days ago, it was caused by sharing pgrps across function jobs.

I can confirm the fix. (for the `grep` crash, that is. The `source`/killpg business still happens)

mqudsi commented:
Yeah, GitHub loses it when rebasing. I just pushed a commit that fixed a regression that wasn't there the last time you tried it (fish high cpu when another program has control of the terminal). I previously also committed a fix for a hang when piping an external command to a function when the external command produces more output than can fit within the OS pipe buffer and that entire chain is in the foreground and buffering was at play, previously we would block on the foreground process with reading from the pipe causing the prior command to block waiting for the pipe to unblock.

I just pushed again with a fix with the killpg warning. It was a bug introduced in 1b1bc28c0ae0fd10e9d3854b1d8cc8125623e1d9 but didn't surface until the changes to job control, fish would detect that the script calling `exec`  hadn't terminated and try killing it too.

mqudsi commented:
Sorry, there was a typo in the killpg fix. Fixed now.

faho commented:
So, @mqudsi, @ridiculousfish, @zanchey: Can we merge this now?

I'd love to, since then we might get some testing in. It's also one of the few things left in the [3.0 milestone](https://github.com/fish-shell/fish-shell/milestone/18).

mqudsi commented:
I guess the point of a pull request is for code review and for testing we're going to have to merge this into master sooner or later (thanks @comfortablynick for trying it out). 

I don't like making tentative changes to master that might need to be reverted, but I guess the need for testing here wins out over aesthetic cleanliness of `master` history. I can rebase, give it a once-over, and merge it into `master` and if there are any problems that bubble up, we can address them or temporarily revert, I suppose.

Reminding myself: I should add test cases for some of the transient bugs I ran into developing this branch. The problem is that I'm still not 100% on how the expect tests work and what type of deviations from the explicitly scripted responses it accepts.

mqudsi commented:
It's done! 😬

JoshCheek commented:
❤️ Ty for this!
